### PR TITLE
[wdspec] Add tests for user activation

### PR DESCRIPTION
**Pull request by OrKoN**
Spec: https://github.com/w3c/webdriver-bidi/pull/496

### CODE REVIEW

**Comment by OrKoN** [comment]
@thiagowfx PTAL

File: webdriver/tests/bidi/script/evaluate/evaluate.py
Comment (jgraham) [inline webdriver/tests/bidi/script/evaluate/evaluate.py:102]: It would be nice to also have a test that depends on the results of having user activation, to ensure that we're correctly enabling the state, not just changing the results of the API call (also because gecko is missing the `navigator.userActivation` API at the moment).

**Review by thiagowfx — APPROVED** [review]

File: webdriver/tests/bidi/script/evaluate/evaluate.py
Comment (OrKoN) [reply webdriver/tests/bidi/script/evaluate/evaluate.py:102]: Added a test with the copy commands.

Comment (OrKoN) [reply webdriver/tests/bidi/script/evaluate/evaluate.py:102]: This is the test we use for Puppeteer although I recognize that the execCommand has been removed an deprecated now? https://developer.mozilla.org/en-US/docs/Web/API/Document/execCommand it works in Chrome and Firefox though. Do you have alternative cases in mind that we could use?

**Review by juliandescottes — COMMENTED** [review]
Thanks, can we add similar tests to call_function?

File: tools/webdriver/webdriver/bidi/modules/script.py
Comment (OrKoN) [reply tools/webdriver/webdriver/bidi/modules/script.py:128]: AFAIK `Optional[bool]` would not allow testing invalid types here?

Comment (juliandescottes) [reply tools/webdriver/webdriver/bidi/modules/script.py:128]: `Optional[bool]` is what we used for `await_promise` before it became mandatory, and we have test for invalid types. So I would be surprised if it didn't work, but I haven't pulled and tested this out.

**Review by juliandescottes — APPROVED** [review]
Thanks!

**Review by thiagowfx — APPROVED** [review]

**Review by whimboo — CHANGES_REQUESTED** [review]
@OrKoN, I have seen that the chrome stability job is failing for the user activation tests. Please have a look at those.

**Comment by OrKoN** [comment]
@whimboo it looks like I need to consume the activation to make tests reliable. Test failures went away but there are still some errors and I am not sure if I have seen those before.

**Comment by whimboo** [comment]
> @whimboo it looks like I need to consume the activation to make tests reliable. Test failures went away but there are still some errors and I am not sure if I have seen those before.

Which failures do you mean in detail? Also shouldn't we have tests including frames as well? Same origin should also get user activation, while cross-origin frames not.

**Comment by whimboo** [comment]
@OrKoN would you mind rebasing your branch? At least the `wpt-firefox-nightly-stability` job issues have been fixed which means the job should pass now.

**Comment by OrKoN** [comment]
@whimboo rebased

**Review by whimboo — APPROVED** [review]
Thanks @OrKoN. Results look much better now. Given the result of the Chrome stability job I'll let you decide if that problem is concerning. The [experimental Chrome results](https://github.com/web-platform-tests/wpt/pull/41217/checks?check_run_id=16571047177) are just fine, so it might not block. Please discuss internally.

**Comment by whimboo** [comment]
> @thiagowfx please review. I think the failures are not related to the change here.

Please note that another PR (https://github.com/web-platform-tests/wpt/pull/41814) as merged yesterday hasn't had any failure for chrome-stability. So I'm not 100% sure here.

**Comment by OrKoN** [comment]
@whimboo I think the job only runs the test files that have been touched. Let me open a dummy change to the same files and see if it makes any difference.

**Comment by whimboo** [comment]
> @whimboo I think the job only runs the test files that have been touched. Let me open a dummy change to the same files and see if it makes any difference.

That's a good idea! The dummy PR is https://github.com/web-platform-tests/wpt/pull/41857.

**Review by thiagowfx — APPROVED** [review]

**Comment by OrKoN** [comment]
@whimboo looks like the addition to invalid.py makes the test qualify as slow for chrome. I will investigate more later.

**Comment by whimboo** [comment]
> @whimboo looks like the addition to invalid.py makes the test qualify as slow for chrome. I will investigate more later.

But that is not the reason for the Python exceptions as seen, and also won't block the landing. There are still 4s left before it reaches the 25s timeout mark.

Your dummy PR also shows the exceptions and as such I think that we can actually merge this PR. Let me know what you think.

**Review by thiagowfx — APPROVED** [review]

**Comment by whimboo** [comment]
@OrKoN do you still think that this is [blocked on the `chrome-dev-stability` job failure](https://github.com/web-platform-tests/wpt/pull/41217/checks?check_run_id=16744376304)?

**Comment by OrKoN** [comment]
@whimboo I have tried splitting the new test into a new file and then it does not seem to cause errors/slowness. I can roll back to the previous version if we want to keep tests in a single file and, then we can force-merge if the job fails. The current version seems to be passing in chrome stability job but fails in the firefox one?

**Review by whimboo — CHANGES_REQUESTED** [review]
Generally it looks fine. See my inline comment for a better place of the invalid test case.

Nevertheless we need the same kind of tests for `script.callFunction` as well.

**Comment by OrKoN** [comment]
@whimboo which inline comment do you mean?

**Comment by whimboo** [comment]
@OrKoN that's interesting. Looks like it got lost. It was about the invalid test that you moved out. Please just get it added to the other invalid tests in `invalid.py`.

**Comment by Ryukung1** [comment]
orkon/user-activation

**Comment by dbaron** [comment]
(apparently I removed multiple review requests when attempting to remove only the one to me, by clicking on my name in the Reviewers UI.  No idea why...)

**Review by thiagowfx — APPROVED** [review]

File: webdriver/tests/bidi/script/call_function/invalid.py
Comment (whimboo) [inline webdriver/tests/bidi/script/call_function/invalid.py:425]: Can you please check why adding this test causes a timeout for Chrome? wpt.fyi doesn't show any result.

File: webdriver/tests/bidi/script/call_function/user_activation.py
Comment (whimboo) [inline webdriver/tests/bidi/script/call_function/user_activation.py:5]:
```suggestion

@pytest.mark.asyncio
```

File: webdriver/tests/bidi/script/evaluate/evaluate.py
Comment (whimboo) [inline webdriver/tests/bidi/script/evaluate/evaluate.py:96]: Similar to the `script.callFunction` tests please move it to its own file.

**Review by whimboo — CHANGES_REQUESTED** [review]

File: webdriver/tests/bidi/script/call_function/invalid.py
Comment (OrKoN) [reply webdriver/tests/bidi/script/call_function/invalid.py:425]: Do you have a link to the failure?

Comment (whimboo) [reply webdriver/tests/bidi/script/call_function/invalid.py:426]: Given that you have pushed to this branch the test are running again, and I cannot find a reference to the wpt.fyi results for an older commit. So lets wait until the tests finished.

Comment (OrKoN) [reply webdriver/tests/bidi/script/call_function/invalid.py:426]: It passes locally with chromium-bidi, so I think it would be related to some WPT infrastructure.

Comment (OrKoN) [reply webdriver/tests/bidi/script/call_function/invalid.py:426]: It finished now, do you still see the same failure? I am not sure where to look for it (I am seeing some infra failures) and the known chrome stability error that says that we approach the slow test limit (due to the number of tests).

File: webdriver/tests/bidi/script/call_function/user_activation.py
Comment (OrKoN) [inline webdriver/tests/bidi/script/call_function/user_activation.py:8]:
```suggestion
async def test_callFunction_userActivation(bidi_session, top_context, user_activation):
```

File: webdriver/tests/bidi/script/call_function/user_activation.py
Comment (OrKoN) [inline webdriver/tests/bidi/script/call_function/user_activation.py:28]:
```suggestion
async def test_callFunction_userActivation_copy(bidi_session, top_context, user_activation):
```

File: webdriver/tests/bidi/script/evaluate/evaluate.py
Comment (OrKoN) [inline webdriver/tests/bidi/script/evaluate/evaluate.py:5]:
```suggestion

@pytest.mark.asyncio
```

File: webdriver/tests/bidi/script/evaluate/evaluate.py
Comment (OrKoN) [inline webdriver/tests/bidi/script/evaluate/evaluate.py:3]:
```suggestion

from ... import recursive_compare
```

File: webdriver/tests/bidi/script/evaluate/user_activation.py
Comment (OrKoN) [inline webdriver/tests/bidi/script/evaluate/user_activation.py:28]:
```suggestion
async def test_userActivation_copy(bidi_session, top_context, user_activation):
```

File: webdriver/tests/bidi/script/evaluate/user_activation.py
Comment (OrKoN) [inline webdriver/tests/bidi/script/evaluate/user_activation.py:8]:
```suggestion
async def test_userActivation(bidi_session, top_context, user_activation):
```

File: webdriver/tests/bidi/script/call_function/invalid.py
Comment (OrKoN) [reply webdriver/tests/bidi/script/call_function/invalid.py:426]: sorry, I have noticed some further inconsistencies in naming so I triggered a re-run again

**Comment by Ms2ger** [comment]
The labels on this PR seem implausible; please review

**Comment by whimboo** [comment]
@OrKoN please have a look at https://github.com/web-platform-tests/wpt/pull/41217/checks?check_run_id=17223989619 and compare it with https://github.com/web-platform-tests/wpt/pull/41217/checks?check_run_id=17223991400. But maybe that's a different issue? Why is only a single test run for Chrome while we have 120 for Firefox?

**Comment by OrKoN** [comment]
@whimboo thanks, yeah, it's odd. The error here says it is "harness timeout" https://wpt.fyi/results/webdriver/tests/bidi/script/call_function?label=pr_head&max-count=1&pr=41217 At the same time evaluate tests seem okay.

**Comment by OrKoN** [comment]
@whimboo it's the same on the main branch https://wpt.fyi/results/webdriver/tests/bidi/script/call_function?label=experimental&label=master&aligned

**Comment by OrKoN** [comment]
I will file an issue for that.

**Comment by OrKoN** [comment]
Filed https://github.com/web-platform-tests/wpt/issues/42220

**Review by whimboo — APPROVED** [review]
Yes, looks good. Thanks!

**Pull request merged**
